toolkit/tools/internal/safechroot/safechroot.go

```diff
@@ -689,3 +699,48 @@ func (c *Chroot) GetMountPoints() []*MountPoint {
 	mountPoints := append([]*MountPoint(nil), c.mountPoints...)
 	return mountPoints
 }
+
+// stopGPGComponents stops gpg-agent and keyboxd if they are running inside the chroot.
+//
+// A GPG agent may have been started while the chroot was in use. Newer versions of "gnupg2" will also start keyboxd.
+// E.g. when installing the azurelinux-repos-shared package, a GPG import occurs. This starts the gpg-agent process inside the chroot.
+// To be able to cleanly exit the setup chroot, we must stop it.
+func (c *Chroot) stopGPGComponents() (err error) {
+	_, err = exec.LookPath("gpgconf")
+	if err != nil {
+		// gpgconf is not installed, so gpg-agent is not running.
+		logger.Log.Debugf("gpgconf is not installed, so gpg-agent is not running: %s", err)
+		return
+	}
+
+	err = c.UnsafeRun(func() (err error) {
+		stdout, stderr, err := shell.Execute("gpgconf", "--list-components")
+
+		logger.Log.Debugf("gpgconf --list-components output:\n%s", stdout)
+
+		if err != nil || stderr != "" {
+			err = fmt.Errorf("failed to list gpg components.\nerr: %s\nstderr: %s\nUnable to check and stop GPG components", err, stderr)
+			return
+		}
+
+		componentsToKill := []string{"gpg-agent", "keyboxd"}
+		// Split --list-components stdout on newline and iterate over each line
+		for _, line := range strings.Split(stdout, "\n") {
```

Comment:
Summary of an offline discussion: let's refactor by extracting the components listing to a separate function.